### ISSUE TITLE
Incorrect resolved type of binary expression

### ISSUE THREAD

**Issue by Shoreray** [javaparser/javaparser#1589]
It seems that the method com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getBinaryTypeConcrete(Node left, Node right, boolean solveLambdas) does not handle type promotion correctly.

For example, when the resolved types of the left and right nodes are byte and char correspondingly, the returned type of the binary expression is byte, instead of the correct type int.

**Comment by Shoreray** [comment]
The issue is not limited to promotion of primitive types. When the operator is "+", the right node of the binary expression is a string but the left node is not, the calculated type of the expression is whatever the type of the left node is.

**Comment by ftomassetti** [comment]
Thanks for reporting this one!

**Comment by matozoid** [comment]
@ftomassetti any hints for where we should look to solve this?

**Comment by ftomassetti** [comment]
The primitive types know about the values to which they can promoted but this is probably not good enough and we should probably change the code that calculates the type of BinaryExpr. There should be a visitor that given an expression returns the type.

**Comment by ftomassetti** [comment]
We should probably change this method in JavaParserFacade:
```java
    protected ResolvedType getBinaryTypeConcrete(Node left, Node right, boolean solveLambdas) {
        ResolvedType leftType = getTypeConcrete(left, solveLambdas);
        ResolvedType rightType = getTypeConcrete(right, solveLambdas);
        if (rightType.isAssignableBy(leftType)) {
            return rightType;
        }
        return leftType;
    }
```

**Comment by ftomassetti** [comment]
Closed via #1732
